### PR TITLE
[GEOS-10012] Support inclusion vendor options to disable Style elements when rendering legends

### PR DESCRIPTION
**Pull request by taba90**
[GEOS-10012](https://osgeo-org.atlassian.net/browse/GEOS-10012) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

Jira ticket https://osgeo-org.atlassian.net/browse/GEOS-10012

- this pr add support for renderingLegend vendor option, enabling the possibility so selectively include elements in getLegendGraphic final output.
- - blocked by https://github.com/geotools/geotools/pull/3447

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.**

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [x] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

File: src/wms/src/main/java/org/geoserver/wms/legendgraphic/BufferedImageLegendGraphicBuilder.java [hunk 1]
```diff
@@ -135,6 +135,7 @@ public BufferedImage buildLegendGraphic(GetLegendGraphicRequest request)
                 throw new NullPointerException("request.getStyle()");
             }
 
+            gt2Style = appliesRenderingSelection(gt2Style);
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/legendgraphic/BufferedImageLegendGraphicBuilder.java:138]:
```suggestion
            gt2Style = applyRenderingSelection(gt2Style);
```

File: doc/en/user/source/styling/sld/extensions/rendering-selection.rst [hunk 2]
```diff
@@ -0,0 +1,305 @@
+.. _rendering_selection:
+
+Rendering Selection
+====================
+
+GeoServer provides two ``VendorOptions`` to define whether a particular element ``Rule``, ``FeatureTypeStyle`` or ``Symbolizer`` should be applied to a ``getLegendGraphic`` output or to a ``getMap`` output.
+The two options are named ``renderingMap`` and ``renderingLegend`` and can take a value of ``true`` or ``false`` eg.:
+
+* - <VendorOption name="renderingMap">false</VendorOption>
```
Comment (aaime) [inline doc/en/user/source/styling/sld/extensions/rendering-selection.rst:9]: To be updated based on the geotools PR changes.

**Review by aaime — CHANGES_REQUESTED** [review]

File: doc/en/user/source/styling/sld/extensions/rendering-selection.rst (hunk 2 above)
Comment (taba90) [reply doc/en/user/source/styling/sld/extensions/rendering-selection.rst]: fixed

File: src/wms/src/main/java/org/geoserver/wms/legendgraphic/BufferedImageLegendGraphicBuilder.java (hunk 1 above)
Comment (taba90) [reply src/wms/src/main/java/org/geoserver/wms/legendgraphic/BufferedImageLegendGraphicBuilder.java]: fixed

File: doc/en/user/source/styling/sld/extensions/rendering-selection.rst [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+GeoServer provides a ``VendorOptions`` to define whether a particular element ``Rule``, ``FeatureTypeStyle`` or ``Symbolizer`` should be applied to a ``getLegendGraphic`` output or to a ``getMap`` output.
+The option is named ``inclusion`` and can take the following values: ``normal`` if no selection will apply, ``mapOnly`` if the SLD element should be used only during maps rendering and ``legendOnly`` if the element will be applied only when rendering legends eg.:
+
+* - <VendorOption name="inclusion">legendOnly</VendorOption>
+
+* - <VendorOption name="inclusion">mapOnly</VendorOption>
+
+If the value is set to ``legendOnly`` the element will be skipped when applying the style to the data to render map.
+If the value is set to ``mapOnly`` the element will be skipped when applying the style to the data to render legend.
+If the value is set to ``normal`` will have the same effect then omitting the VendorOption: the SLD element will be used for both map and legend.
```
Comment (aaime) [inline doc/en/user/source/styling/sld/extensions/rendering-selection.rst:15]: Looks like a great use case for a bullet list.

```suggestion
Valid values are:
-  ``legendOnly`` the element will be skipped when applying the style to the data to render map.
- ``mapOnly`` the element will be skipped when applying the style to the data to render legend.
- ``normal`` will have the same effect then omitting the VendorOption: the SLD element will be used for both map and legend (same effect as not specifying the vendor option).
```

File: doc/en/user/source/styling/sld/extensions/rendering-selection.rst [hunk 4]
```diff
@@ -0,0 +1,306 @@
+.. _rendering_selection:
+
+Rendering Selection
+====================
+
+GeoServer provides a ``VendorOptions`` to define whether a particular element ``Rule``, ``FeatureTypeStyle`` or ``Symbolizer`` should be applied to a ``getLegendGraphic`` output or to a ``getMap`` output.
+The option is named ``inclusion`` and can take the following values: ``normal`` if no selection will apply, ``mapOnly`` if the SLD element should be used only during maps rendering and ``legendOnly`` if the element will be applied only when rendering legends eg.:
```
Comment (aaime) [inline doc/en/user/source/styling/sld/extensions/rendering-selection.rst:7]: The definition of what values do was specified twice.

```suggestion
The vendor option is named ``inclusion``, e.g.:
```

File: doc/en/user/source/styling/sld/extensions/rendering-selection.rst [hunk 5]
```diff
@@ -0,0 +1,306 @@
+.. _rendering_selection:
+
+Rendering Selection
+====================
+
+GeoServer provides a ``VendorOptions`` to define whether a particular element ``Rule``, ``FeatureTypeStyle`` or ``Symbolizer`` should be applied to a ``getLegendGraphic`` output or to a ``getMap`` output.
```
Comment (aaime) [inline doc/en/user/source/styling/sld/extensions/rendering-selection.rst:6]:
```suggestion
GeoServer provides a ``VendorOptions`` to define whether a particular element ``Rule``, ``FeatureTypeStyle`` or ``Symbolizer`` should be applied to a ``getLegendGraphic`` output or to a ``getMap`` output.
This allows the generate legends from the SLD that can better looking and more expressive, without the underlying complexity of the actual rendered style. Other systems have a dedicated language to build legends instead. The advantage of using the same language is that dynamic behaviors, like rule removal based on the area being rendered, can be easily retained.
```

**Review by aaime — CHANGES_REQUESTED** [review]
Code changes look good, proposed small changes in the docs.

File: doc/en/user/source/styling/sld/extensions/rendering-selection.rst (hunk 3 above)
Comment (taba90) [reply doc/en/user/source/styling/sld/extensions/rendering-selection.rst]: fixed

File: doc/en/user/source/styling/sld/extensions/rendering-selection.rst (hunk 4 above)
Comment (taba90) [reply doc/en/user/source/styling/sld/extensions/rendering-selection.rst]: fixed

File: doc/en/user/source/styling/sld/extensions/rendering-selection.rst (hunk 5 above)
Comment (taba90) [reply doc/en/user/source/styling/sld/extensions/rendering-selection.rst]: fixed

**Comment by taba90** [comment]
@thanks for your review @aaime your feedback has been applied

**Review by aaime — APPROVED** [review]

**Comment by aaime** [comment]
Thanks, merged!

**Pull request merged**